### PR TITLE
http: allow passing in pre-existing connection bev

### PR DESCRIPTION
**Pull request by theoden8**
resubmit of

- #479

note that this is different from
- #907
- #322

as it allows passing file descriptors

attempting
- https://github.com/libevent/libevent/issues/1471#issuecomment-2367518127

pending examples

### CODE REVIEW

**Comment by theoden8** [comment]
Use after free of `req` when `con_outgoing` is true:

https://github.com/libevent/libevent/blob/970a258bc6ee08abf4cf81540570f0b8eca79449/http.c#L931-L974

File: include/event2/http.h [hunk 1]
```diff
@@ -674,7 +674,9 @@ enum evhttp_request_kind { EVHTTP_REQUEST, EVHTTP_RESPONSE };
  *     specified host name resolution will block.
  * @param bev a bufferevent to use for connecting to the server; if NULL, a
  *     socket-based bufferevent will be created.  This bufferevent will be freed
- *     when the connection closes.  It must have no fd set on it.
+ *     when the connection closes.  If a fd is already set on the bufferevent,
```
Comment (azat) [inline include/event2/http.h:677]: Such API looks too tricky, I guess it will be better a separate method for this

Comment (theoden8) [reply include/event2/http.h]: Added a function `evhttp_connection_base_bufferevent_set_new` into the interface

**Comment by theoden8** [comment]
@azat ready for review

**Comment by theoden8** [comment]
The TBSAN/ASAN/UBSAN issues are reproducible in the current `master` at 528fbed18426cbf0810e085cce7b80ab4654bce0 on the tests that do not use preexisting connection fd

File: include/event2/http.h [hunk 2]
```diff
@@ -699,6 +699,24 @@ EVENT2_EXPORT_SYMBOL
...
+ *     specified host name resolution will block.
+ * @param bev a bufferevent to use for connecting to the server. A fd
+ *     is already set on the bufferevent, it will be assumed that this
+ *     connection is already open and ready to send requests.
+ * @return an evhttp_connection object that can be used for making requests or
+ *   NULL on error
+ */
+EVENT2_EXPORT_SYMBOL
+struct evhttp_connection *
+evhttp_connection_base_bufferevent_set_new(struct event_base *base, struct evdns_base *dnsbase, struct bufferevent* bev);
```
Comment (azat) [inline include/event2/http.h:718]:
```suggestion
evhttp_connection_base_bufferevent_reuse_new(struct event_base *base, struct evdns_base *dnsbase, struct bufferevent* bev);
```

Maybe reuse will be a better name?

File: http.c [hunk 3]
```diff
@@ -2531,6 +2531,19 @@ evhttp_connection_new(const char *address, ev_uint16_t port)
...
+ * Assume that this is an already-open connection that we
+ * can start sending requests on.
+ */
+static int
+evhttp_connection_set_existing_(struct evhttp_connection *evcon, struct bufferevent* bev)
+{
+	evcon->state = EVCON_IDLE;
+	evcon->flags |= EVHTTP_CON_OUTGOING;
+	return 0;
+
```
Comment (azat) [inline http.c:2544]:
```suggestion
```

File: test/regress_http.c [hunk 4]
```diff
@@ -2333,7 +2365,14 @@ static void http_unix_socket_test(void *arg)
 	uri = evhttp_uri_parse_with_flags(uri_loc, EVHTTP_URI_UNIX_SOCKET);
 	tt_assert(uri);
 
-	evcon = evhttp_connection_base_bufferevent_unix_new(data->base, NULL, evhttp_uri_get_unixsocket(uri));
+	if (preexisting) {
+		client_fd = http_connect_unixsocket(tmp_sock_path);
+		tt_fd_op(client_fd,!=,EVUTIL_INVALID_SOCKET);
```
Comment (azat) [inline test/regress_http.c:2370]:
```suggestion
		tt_fd_op(client_fd, !=, EVUTIL_INVALID_SOCKET);
```

File: test/regress_http.c [hunk 5]
```diff
@@ -2333,7 +2365,14 @@ static void http_unix_socket_test(void *arg)
...
+		bev = create_bev(data->base, client_fd, 0x00, BEV_OPT_CLOSE_ON_FREE);
+		evcon = evhttp_connection_base_bufferevent_set_new(data->base, NULL, bev);
```
Comment (azat) [inline test/regress_http.c:2372]:
```suggestion
		evcon = evhttp_connection_base_bufferevent_reuse_new(data->base, NULL, bev);
```

File: test/regress_http.c [hunk 6]
```diff
@@ -2351,6 +2390,10 @@ static void http_unix_socket_test(void *arg)
 
 	event_base_dispatch(data->base);
 
+	if (preexisting) {
+		tt_fd_op(bufferevent_getfd(bev),==,client_fd);
```
Comment (azat) [inline test/regress_http.c:2394]:
```suggestion
		tt_fd_op(bufferevent_getfd(bev), ==, client_fd);
```

File: test/regress_http.c [hunk 7]
```diff
@@ -4383,30 +4438,46 @@ http_make_web_server(evutil_socket_t fd, short what, void *arg)
...
-	bufferevent_mbedtls_set_allow_dirty_shutdown(bev, dirty);
+	if(ssl == HTTP_MBEDTLS) {
+		bufferevent_mbedtls_set_allow_dirty_shutdown(bev, dirty);
+	}
 #endif
 
-	evcon = evhttp_connection_base_bufferevent_new(
-		data->base, NULL, bev, "127.0.0.1", hs.port);
+	if (preexisting) {
+		evcon = evhttp_connection_base_bufferevent_set_new(data->base, NULL, bev);
```
Comment (azat) [inline test/regress_http.c:4472]:
```suggestion
		evcon = evhttp_connection_base_bufferevent_reuse_new(data->base, NULL, bev);
```

File: test/regress_http.c [hunk 8]
```diff
@@ -4383,30 +4438,46 @@ http_make_web_server(evutil_socket_t fd, short what, void *arg)
...
 	struct evhttp *http = http_setup(&hs.port, data->base, ssl);
+	evutil_socket_t client_fd = EVUTIL_INVALID_SOCKET;
 
 	exit_base = data->base;
 	test_ok = 0;
 
-	bev = create_bev(data->base, -1, ssl, BEV_OPT_CLOSE_ON_FREE);
+	if (preexisting) {
+		client_fd = http_connect("127.0.0.1", hs.port);
+		tt_fd_op(client_fd,!=,EVUTIL_INVALID_SOCKET);
```
Comment (azat) [inline test/regress_http.c:4456]:
```suggestion
		tt_fd_op(client_fd, !=, EVUTIL_INVALID_SOCKET);
```

File: test/regress_http.c [hunk 9]
```diff
@@ -4383,30 +4438,46 @@ http_make_web_server(evutil_socket_t fd, short what, void *arg)
...
+	}
+	bev = create_bev(data->base, client_fd, ssl, BEV_OPT_CLOSE_ON_FREE);
+	tt_fd_op(bufferevent_getfd(bev),==,client_fd);
```
Comment (azat) [inline test/regress_http.c:4459]:
```suggestion
	tt_fd_op(bufferevent_getfd(bev), ==, client_fd);
```

File: test/regress_http.c [hunk 10]
```diff
@@ -4383,30 +4438,46 @@ http_make_web_server(evutil_socket_t fd, short what, void *arg)
...
 #ifdef EVENT__HAVE_OPENSSL
-	bufferevent_openssl_set_allow_dirty_shutdown(bev, dirty);
+	if(ssl == HTTP_OPENSSL) {
```
Comment (azat) [inline test/regress_http.c:4461]:
```suggestion
	if (ssl == HTTP_OPENSSL) {
```

File: test/regress_http.c [hunk 11]
```diff
@@ -4383,30 +4438,46 @@ http_make_web_server(evutil_socket_t fd, short what, void *arg)
...
+		bufferevent_openssl_set_allow_dirty_shutdown(bev, dirty);
+	}
 #endif
 #ifdef EVENT__HAVE_MBEDTLS
-	bufferevent_mbedtls_set_allow_dirty_shutdown(bev, dirty);
+	if(ssl == HTTP_MBEDTLS) {
```
Comment (azat) [inline test/regress_http.c:4466]:
```suggestion
	if (ssl == HTTP_MBEDTLS) {
```

**Review by azat — COMMENTED** [review]
LGTM

**Review by azat — APPROVED** [review]

File: http.c [hunk 12]
```diff
@@ -2585,6 +2597,35 @@ evhttp_connection_new_(struct event_base *base, struct bufferevent* bev)
 	return (NULL);
 }
 
+struct evhttp_connection *
+evhttp_connection_base_bufferevent_set_new(struct event_base *base, struct evdns_base *dnsbase, struct bufferevent* bev)
```
Comment (azat) [inline http.c:2601]:
```suggestion
evhttp_connection_base_bufferevent_reuse_new(struct event_base *base, struct evdns_base *dnsbase, struct bufferevent* bev)
```

**Pull request merged**